### PR TITLE
SiteNow Dispatch Module

### LINKED ISSUE #3302: Signup for Dispatch Communications
# Story

As a SiteNow site owner, I would like the ability to have users sign up to dispatch communications from my website so that users can receive newsletters and other dispatch communications.

# Additional Info

Charlie Taylor would like this functionality on diversity.uiowa.edu.

Our SiteNow site is probably a good test site to use while working on this.

~`Michael Alberhasky has some initial work on the Dispatch side of this and will follow up in a couple weeks.~ This work had been done... https://apps.its.uiowa.edu/dispatch/api-ref#/SubscriptionLists

# Proposed solution

Create a uiowa_dispatch module which has two things initially:

- A configuration page for webmaster to add at least their client API key. More configuration may be needed.
- A Signup Form custom block that takes configuration for a communication/population. Use the API key to query new API endpoint and return embedded signup form.

Because the form can change arbitrarily on Dispatch side, this was preferred method rather than constructing ourselves. AIS agreed to style after UIDS and can add/remove classes as needed.

This module could hold more Dispatch integrations in the future.

### PR DESCRIPTION
**Pull request by GaryRidgway**
Resolves #3302 

## To test

Sync down a site
Add the block
enter a non-uiowa email and first & last name
check the ITS email that something got sent through
[Verify here](https://apps.its.uiowa.edu/dispatch/populations/718296392/members) that the email got added to the list.
Delete the example email that got added so it does not clutter the list
?????
Profit.

### \*\*hint\*\*
Ask Alan for an API key

### CODE REVIEW

**Comment by joewhitsitt** [comment]
```
FILE: ...les/custom/sitenow_dispatch/src/Plugin/Block/SignUpFormBlock.php
----------------------------------------------------------------------
FOUND 12 ERRORS AND 2 WARNINGS AFFECTING 14 LINES
----------------------------------------------------------------------
   7 | WARNING | [x] Unused use statement
  11 | WARNING | [x] Unused use statement
  59 | ERROR   | [x] Missing function doc comment
  60 | ERROR   | [x] Opening brace should be on the same line as the
     |         |     declaration
  64 | ERROR   | [x] Missing function doc comment
  65 | ERROR   | [x] Opening brace should be on the same line as the
     |         |     declaration
  71 | ERROR   | [x] Missing function doc comment
  72 | ERROR   | [x] Opening brace should be on the same line as the
     |         |     declaration
  80 | ERROR   | [x] Expected 1 space after "="; 0 found
  96 | ERROR   | [x] Missing function doc comment
 112 | ERROR   | [x] Missing function doc comment
 113 | ERROR   | [x] Opening brace should be on the same line as the
     |         |     declaration
 116 | ERROR   | [x] Expected 1 blank line after function; 0 found
 117 | ERROR   | [x] The closing brace for the class must have an
     |         |     empty line before it
----------------------------------------------------------------------
PHPCBF CAN FIX THE 14 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```

**Comment by joewhitsitt** [comment]
Haven't tested yet, but `core.entity_view_display.node.page.default.yml` and permissions are split by several sites so if you can't add this block through one of the existing categories and existing permissions, you are going to have to update all of those other split files with this.

File: docroot/modules/custom/sitenow_dispatch/sitenow_dispatch.install [hunk 1]
```diff
@@ -0,0 +1,96 @@
+<?php
```
Comment (richardbporter) [inline docroot/modules/custom/sitenow_dispatch/sitenow_dispatch.install]: I don't think this file is needed.

File: docroot/modules/custom/sitenow_dispatch/sitenow_dispatch.libraries.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Custom module library for general purposes.
```
Comment (richardbporter) [inline docroot/modules/custom/sitenow_dispatch/sitenow_dispatch.libraries.yml]: Nor is this.

File: docroot/modules/custom/sitenow_dispatch/sitenow_dispatch.routing.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+sitenow_dispatch.example:
```
Comment (richardbporter) [inline docroot/modules/custom/sitenow_dispatch/sitenow_dispatch.routing.yml]: This route can be deleted and the controller.

**Review by richardbporter — CHANGES_REQUESTED** [review]

**Comment by richardbporter** [comment]
I wonder if we should add some error handling to this, maybe another message if the user is already subscribed (409 response) or if something goes wrong?

**Comment by richardbporter** [comment]
I get an error trying to add the block to the page when no API key exists. This could be confusing to people. We should probably check to see if an API key has been entered and output a message if not linking them to the settings form.

**Comment by richardbporter** [comment]
We should add some help text to admin/config/system/sitenow-dispatch with some links to the Dispatch documentation.

**Comment by GaryRidgway** [comment]
Things still to do

- [x] Add user checking to not subscribe a person again if they are subscribed.
- [x] check for API key before adding block, link to settings form if not filled out.
- [x] add help text to dispatch settings form linking to dispatch docs
- [x] provide better success message instead of "You've been Aardvarked!"
- [x] permissions????

File: docroot/modules/custom/sitenow_dispatch/src/Form/SettingsForm.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $form['description_text'] = [
+      '#markup' => '<p>This API key is for the Dispatch service, references are listed below for ease of use.</p><p><a href="https://apps.its.uiowa.edu/dispatch/help">Dispatch documentation</a></p><p><a href="https://apps.its.uiowa.edu/dispatch/api-ref">API reference</a></p>',
+    ];
+    $form['API_key'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('API key'),
+      '#default_value' => $this->config('sitenow_dispatch.settings')->get('API_key'),
```
Comment (richardbporter) [inline docroot/modules/custom/sitenow_dispatch/src/Form/SettingsForm.php]: Nitpick: lower case the config key?

**Comment by richardbporter** [comment]
> Haven't tested yet, but `core.entity_view_display.node.page.default.yml` and permissions are split by several sites so if you can't add this block through one of the existing categories and existing permissions, you are going to have to update all of those other split files with this.

I think we forgot to update split `core.entity_view_display.node.page.default.yml` files.

File: docroot/modules/custom/sitenow_dispatch/src/Form/SettingsForm.php (hunk 4 above)
Comment (GaryRidgway) [reply docroot/modules/custom/sitenow_dispatch/src/Form/SettingsForm.php]: you watched me put this in there! you just wanted me to do more work >:(

**Comment by GaryRidgway** [comment]
> > Haven't tested yet, but `core.entity_view_display.node.page.default.yml` and permissions are split by several sites so if you can't add this block through one of the existing categories and existing permissions, you are going to have to update all of those other split files with this.
> 
> I think we forgot to update split `core.entity_view_display.node.page.default.yml` files.

isnt this just a re-order?

**Comment by richardbporter** [comment]
I think referring non-developers to the API docs is just going to lead to confusion:

Maybe we should omit that?

**Comment by GaryRidgway** [comment]
> I think referring non-developers to the API docs is just going to lead to confusion:
> 
> Maybe we should omit that?

Idk, is sending them to the documentation any better? will they actually read that?

**Comment by richardbporter** [comment]
One last thing we probably want to do here is spam prevention. @joewhitsitt do you know if custom forms get spam fields auto-added or do we have examples of how to do that?

**Comment by richardbporter** [comment]
I've tested this functionality and it works well. 👍🏻 Might defer approval for someone else on the team.

**Comment by joewhitsitt** [comment]
we have honeypot that can be added to custom forms: https://www.drupal.org/project/honeypot

you can also set up antibot to cover the form: admin/config/user-interface/antibot

**Comment by GaryRidgway** [comment]
> admin/config/user-interface/antibot

added antibot to it, thanks Joe!

**Review by richardbporter — APPROVED** [review]

**Pull request merged**